### ISSUE TITLE
InfiniteScroll snap script

### ISSUE THREAD

**Issue by SimonDarksideJ** [Unity-UI-Extensions/com.unity.uiextensions#168]
<sup>Issue created by Tomek S as [Bitbucket Issue #&#8203;132](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132) on 2017.04.06 15:05.</sup>
Hi, 
I did the infinite scroll script which is currently part of this UI package, have been recently asked few times for a snap script that will work along with this infinity scroll, I have the script that works fine right off the bat with the infinitescroll script, providing slow/quick swipe options as well as left/right button support., the question is: if is it a good idea to have another scroll snap script in the package or, maybe the current scroll snap should be adjusted to support the infinite scroll? or maybe the script should be just in WIki so whoever needs it can grab it from there. What you think ?

Here is the demo video (also did a small for adding a very quick button pointer down animation (included in the video)) https://www.youtube.com/watch?v=9jDkvmrMGBU

Let me know @ddreaper what you think. I can post the code here if needed.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.04.06 18:47](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-35674524), @SimonDarksideJ commented:</sup>
If you could submit a PR that would be better. We are using GIT now so it should be easier.
Else just put it on Pastebin, drop the link here and I'll merge it manually.

Planning on focusing on the UI project next week so it's good timing

Thanks for the support

Yeah, couldn't use this with the existing scroll snaps as they have their own implementation.  As I love variety, we could always use more :D

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.04.08 07:22](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-35834564), Tomek S commented:</sup>
Ok here is the https://pastebin.com/NXYu37jC button scale, instead of using OnClick (which is too late) is uses pointer down and up events (demo is in the end part of the video above)

and Here infinite scroll snap https://pastebin.com/0hBRTyCC. Features:
* - snapping to position that works with infinite scroll script.
* - slow scroll that centers to the scroll element that is the closest to the center of the viewport
* - quick scroll (quick swipe) to browse to next/previous element, for the better customization I use both, swipe length and time. 
* - easy to plug left/right button
* - has and Unity (int) event whenever the scroll has ended snapping (this is not in the video, added it afterwards)

Remarks:
* - only works for horizontal scroll, would need to add option for vertical but currently I don't have time. (someone can get to it, just need to change the horizontalNormalizedPosition to vertical in coroutines, and for swipe maybe use screen height instead)
* - to retrieve the absolute index of the scroll use GetRelativePage method.
- scroll script need to have flag init by user active in order to work.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.06.22 01:00](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-37757824), @SimonDarksideJ commented:</sup>
Tested the code initially and it didn't snap?
Not entirely sure why as it looks sound. 

Replication Steps
1. Create ScrollRect with content
2. Added infinite scroll (tested to ensure it was scrolling)
3. Added Inf scroll snap - Didn't snap?

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.06.22 06:24](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-37760810), Tomek S commented:</sup>
Well only thing that comes to my mind is the active by user flag on infinite scroll, (this thought should be done automatically too)? This script clearly has issues like no support or vertical scrolling, I will re-test and update it when I will get some time.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.07.08 09:16](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-38100379), @SimonDarksideJ commented:</sup>
Any update @TomekSzelki ?  Nearing the end of this update cycle and would like to have this included.
Just so i know if I need to focus on it or not.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.07.09 10:50](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-38108791), Tomek S commented:</sup>
Hi
I have had few reports of this snap not working. Seems like it just work
with my initial setup of childs of the scroll and when this changes the
snap stops working. It requires a full redo with rethinking the approach
for which sadly I dont have time. So for the time being lets it be a hint
how one would approach this problem. Unless some1 else wants to take over.

Regards

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.07.09 11:11](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-38108942), @SimonDarksideJ commented:</sup>
No problem @TomekSzelki Will put it on hold unless I find some miracle time for this update.  Let me know if you make any improvements.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.18 15:45](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-40614311), jon commented:</sup>
Hi, your UI-infinite scroll does not work. I have my canvas in world space this is what i get:  Assertion failed: Screen position out of view frustum (screen pos 373.874268, 302.255981, -286745.218750) (Camera rect 0 0 852 511)

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.19 12:59](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-40642132), @SimonDarksideJ commented:</sup>
Just tested the UI Infinite scroll does work as expected on a Worldspace canvas, even in 2017.2.0.
Although I did see some abhorrent behaviour when I used the default Unity ScrollView implementation.

So if you are using the UI Infinite Scroll, set up the ScrollRect manually (Parent, content and Layout of children)
I'm also noticing a change to the Layout Groups in 2017, although Unity already recommended avoiding using them.

We don't have any Unit Test code within the Unity UI Extensions, so that error is either coming from Unity or your own code possibly.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.19 14:01](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-40643603), jon commented:</sup>
I don't think this script is designed to work with content generated at runtime it must be why it does not work properly for me. I load texture from streaming asset - > add then into the canvas content -> add a button component for each. which is done from another script. then with ui infinite scroll -> I convert it to work with the hand script from steamVR but fails at even setting the canvas inifinitly. even with a button event listener.
.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.19 16:51](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-40648592), @SimonDarksideJ commented:</sup>
There is an option on the component called "User enabled", which means the script delays alignment of the content until you call it programmatically.

If you are using dynamic content, you will need to use this and call the functions on the script when you change the content.
Hope that helps.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.28 15:48](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-40858126), Dan Bridge commented:</sup>
Hi,

Just checking if the UI_InfiniteScroll component works with HorizontalScrollSnap ok.

I've got "unexpected" behaviour ;) In that the scroll content is infinite but I can't scroll beyond the initial elements in the HSS i.e. I populate the HSS programmatically, calling AddChild to add my scroll items, then call Init on the UI_InfiniteScroll component. Visually all looks well but it wont scroll past the 1st and last items, although I can see the last item to the left of the first. 

Just wondered if there's something else that's needed beyond a vanilla setup of the the UI>Extensions>HSS and adding the UI_InfiniteScroll

p.s. Page Changed Event fires fine for elements until you try and go beyond the initial list i.e. going to n+1 or n-1 of n scroll elements

Any pointers gratefully recieved!

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.30 09:19](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-40883360), @SimonDarksideJ commented:</sup>
If you check Gitter, I posted an experimental update to the ScrollSnap base script, which allows the UIInfinite scroll script to be attached to the HSS/VSS components.  Not uploaded it to source yet as it's not passed my rigorous testing as yet, buttons don't work and a few other tweaks needed.

On the whole though, it works as expected with swipe and events.

**Comment by SimonDarksideJ** [comment]
<sup>On [2017.10.30 13:32](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-40890000), Dan Bridge commented:</sup>
Thanks Simon, I'll have a look now

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.03.17 01:03](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-43974681), Nimrod Ben Simon commented:</sup>
anything new on this? is there a script that works?

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.03.22 18:44](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-44113717), @SimonDarksideJ commented:</sup>
I have checked the updated ScrollSnapBase script in to master a while back, so if you clone / download the repo, you can try out the patched version.
Just be aware, it is still experimental and doesn't support working with the buttons yet.

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.05.08 14:41](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-45297713), Bastiaan Hofsteenge commented:</sup>
I believe we had similar issues as described by Dan Bridge. Scrolling forwards works quite well, scrolling back beyond the initial first item introduces strange behaviour. 
The cause for this was that _infiniteWindow was set to 1 too early while it should be still 0. 
We changed this (in ScrollSnapBase.cs):


```c#
               //Work out which infinite window we are in
                    _infiniteWindow = value / _screensContainer.childCount;
                    //Invert the value if negative and differentiate from Window 0
                    _infiniteWindow = value < 0 ? (-_infiniteWindow) + 1 : _infiniteWindow;
```
Into this:

```c#
                    float infWindow = (float)value / (float)_screensContainer.childCount;          
                    if (infWindow < 0)
                    {
                        _infiniteWindow = (int)(Math.Floor(infWindow));                        
                    }else
                    {
                        _infiniteWindow = value / _screensContainer.childCount;
                    }
                    _infiniteWindow = value < 0 ? (-_infiniteWindow) : _infiniteWindow;
```
Now all works fine as long as we don't use fast swipe or hard swipe, then some crazyness returns.

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.05.08 16:18](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-45301355), @SimonDarksideJ commented:</sup>
Thanks @viemr_Bastiaan , will check your fix out.

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.05.31 15:33](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-45854626), @SimonDarksideJ commented:</sup>
So I applied that fix and while it works for Horizontal, it prevents the vertical scrollsnap from working completely :S So I'll will need to do further investigation.

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.05.31 16:38](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-45856108), @SimonDarksideJ commented:</sup>
Darn it, found the problem with vertical and it was my fault. there was an additional offset check to support infinite and I only added it to the horizontal axis. Applied to vertical and it's all fine now working with this fix.  Thanks everyone for supporting this and to @viemr_Bastiaan for the fix

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.05.31 16:42](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-45856204), @SimonDarksideJ commented:</sup>
Applied for from @viemr_Bastiaan and resolved outstanding issue with the Vertical ScrollSnap when combined with the UI_InfiniteScroll extension.

resolves #132


Referenced by f139f938fdfb

**Comment by SimonDarksideJ** [comment]
<sup>On 2018.05.31 16:42 @SimonDarksideJ modified issue:</sup>
**status** changed `new` → `resolved`

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.07.20 15:13](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-46848560), Matt commented:</sup>
@ddreaper  this issue says resolved but I am still unable to use the infinite scroll with horizontal snap. Has this been updated or did I just miss something? Also is there any demo or documentation for using infinite scroll with HSS?

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.07.25 13:32](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-46930931), @SimonDarksideJ commented:</sup>
It was updates in source, are you using the source version.  I applied some fixes and asked for replies. When I got none it was assumed to be resolved.
I've added notes to your new issue.

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.07.25 13:41](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-46931151), Matt commented:</sup>
I grabbed the source code and it didn't appear to have all of the latest changes, so I just pulled in the different code from the other branch.

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.07.25 14:04](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-46931690), @SimonDarksideJ commented:</sup>
That's curious, apart from the package references, there shouldn't be any difference between the two. Thanks for the heads up, I'll investigate further

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.08.20 20:52](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-47358292), Patryk Sladek commented:</sup>
A quick fix for the next button 
```c#

public void NextScreen()
        {
            if (_currentPage < _screens - 1 || _isInfinate)
            {
                if (!_lerp) StartScreenChange();

                _lerp = true;
                if (_isInfinate)
                {
                    CurrentPage = GetPageforPosition(_screensContainer.localPosition) +1;
                }
                else
                {
                    CurrentPage = _currentPage + 1;
                }

                GetPositionforPage(_currentPage, ref _lerp_target);
                ScreenChange();
            }
```

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.08.22 18:23](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-47397197), Matt commented:</sup>
Thank you so much @Sladzio your fix works perfect. And to fix the previous button you can just change it to: 


```c#

public void PrevScreen()
        {
            if (_currentPage > 0 || _isInfinate)
            {
                if (!_lerp) StartScreenChange();

                _lerp = true;
                if (_isInfinate)
                {
                    CurrentPage = GetPageforPosition(_screensContainer.localPosition) -1;
                }
                else
                {
                    CurrentPage = _currentPage - 1;
                }

                GetPositionforPage(_currentPage, ref _lerp_target);
                ScreenChange();
            }
}
```

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.10.02 22:03](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-48129445), @SimonDarksideJ commented:</sup>
Thanks for the tip @cloud_canvas . I'll get right on that and include it in my next batch of tests

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.10.03 20:08](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-48151841), Nick Suda commented:</sup>
It seems that a similar logic would need to be applied to `GoToScreen()`. Incrementing and decrementing as if successively paginating is one thing, but I'm having a big of a brainfuck trying to figure out how `GoToScreen()` should work. Ideally, it would take in the index number of the unwrapped collection of elements. (Like @danbridge , I'm procedurally populating the HSS and then calling `UIInfiniteScroll.Init()`.) It's just that `UIInfiniteScroll` is set up to constantly adjust your sibling index so that your selection is in the middle.

In other words, if I have a list of 10 elements, part of my initialization process is to wiggle the ScrollRect component by some imperceptible amount so that `UIInfiniteScroll` shuffles the sibling order of the ScrollRect's contents. Item 0 becomes item 4 in a 10-item collection. Any call to move to a page beyond `NextScreen()` and `PreviousScreen()` is going to create an infinite loop.

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.10.03 22:39](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-48153970), Nick Suda commented:</sup>
Yeah, I think the merger of functionality between these two components still has a bit of a way to go. If you could explain how we should optimally be using `UIInfiniteScroll.InitByUser`, `UIInfiniteScroll.Init()` and `ScrollSnapBase.GoToScreen(int newIndex)` when these two components are being used together, that would clear up a lot of confusion. As it is, I can't really get these to play along 100% correctly in the context of procedurally instantiated child objects. This combo is the only thing out there that does this exact feature set (Snapping with nicely controllable lerping and threshold definition, infinite scrolling, increment/decrement by button, pagination, page jumping by script) and it's so close to working altogether.

**Comment by SimonDarksideJ** [comment]
<sup>On [2018.12.17 13:06](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-49504258), Joseph Roney commented:</sup>
Hey everyone, I've been trying to implement Infinite Scrolling with Vertical Scroll Snap (using version 2.0 of the UnityPackage) by adding the UI_InfiniteScroll script/component to the same GO as VerticalScrollSnap, but looks like it doesn't work? I only found this issue thread after that, so just wanted to check, have the updated scripts (ScrollSnapBase.cs, I think?) been updated in the unitypackage file already? Tried it, didn't work.

I also then tried replacing the entire Scripts folder from the unitypackage with the current Scripts folder in the Source (also had to replace Editor\UIExtensionsMenuOptions.cs with the newer version from the Source, since it was giving me an error). Still doesn't seem to be working though.

What happens is that VerticalScrollSnap works fine on its own, but as soon as I add the InfiniteScroll component and hit play, the entire contents of the GO just disappears, so there's pretty much nothing to scroll with. Am I doing something wrong, or has the source just not been updated yet?

Not sure if this was the right place to put this, but it seemed like everyone was on the topic anyway. Forgive me if this is indeed the wrong place for questions like this.
Any help would be greatly appreciated, thanks!

EDIT: To be clearer, I don't need any of the other features like the scroll buttons and stuff to work, at least not yet. Just trying to implement plain Infinite Vertical Scroll Snapping. Thanks again!

**Comment by SimonDarksideJ** [comment]
<sup>On [2020.06.30 12:50](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-57997997), @SimonDarksideJ commented:</sup>
It’s been a while sadly.  But this fix will be included in the upcoming 2.2 \(2020.1\) release

**Comment by SimonDarksideJ** [comment]
<sup>On [2020.06.30 12:56](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-57998098), @SimonDarksideJ commented:</sup>
fix: HSS/VSS ScrollSnap button fixes when infinite control applied

Resolves #132


Referenced by d393830a1d85

**Comment by SimonDarksideJ** [comment]
<sup>On [2020.10.14 12:30](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-59044732), Dmitriy Margulis commented:</sup>
Oh, I really don’t understand how to combine infinite scroll\(at least horizontal\) with snapping. I saw you added this functionality into Horizontal Scroll Snap script in version 1.2, but I didn’t find any tutorial about it as well as methods/properties in scripts. Can you just leave a link here to necessary files as it seems the issue had been resolved a long time ago.

**Comment by SimonDarksideJ** [comment]
<sup>On [2020.10.14 15:14](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-59047087), @SimonDarksideJ commented:</sup>
You simply add the Infinitescroll script to either the Horizontal or Vertical Scroll snap controls, it should be that simple

**Comment by SimonDarksideJ** [comment]
<sup>On [2020.10.14 20:22](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-59049650), Dmitriy Margulis commented:</sup>
Simon, didn’t expect you to answer me so soon. First of all I’d like to thank all of you guys for the saved tons of hours of developing time! Yep, what I found out so far is that component infinite scroll works nicely only with Horizontal Scroll Snap component, which, by default sets size of each element to fill an entire viewport, only one at a time. What I need to achieve is tree elements layout with central one snapping. Exactly like this : [https://www.youtube.com/watch?v=oJ-yqiCYSTk](https://www.youtube.com/watch?v=oJ-yqiCYSTk) \(btw author of this video didn’t left any cue how does it work\)  
Other components like ContentSnapScrollHorizontal or Scroll Snap\(multiple\) don’t have a compatibility with this component - new items spawn before and after content object, but as you release, layout snaps back to the beginning/end. Btw I decided to use just Scroll snap for my purposes without infinite scroll  
UPD. I’ve seen magnetic infinite scroll proposal in the issues recently. May be this is what am I looking for? :slight_smile:

**Comment by SimonDarksideJ** [comment]
<sup>On [2020.10.15 12:35](https://bitbucket.org/UnityUIExtensions/unity-ui-extensions/issues/132#comment-59055605), @SimonDarksideJ commented:</sup>
We aim to please @margulis.d

If there are new things you are looking for, please raise an issue of type proposal for each item and we’ll investigate.

Almost finished with the Magnetic infinite scroll
